File: stack.cpp
#include "stack.h"
#include <iostream>

void CreateStack(Stack* stack) {
    stack -> top = nullptr;
    stack -> size = 0;
}

void SPush(Stack* stack, string data) {
    SNode* newNode = new SNode();
    newNode -> data = data;
    newNode -> next = stack -> top;
    stack -> top = newNode;
    stack -> size++;
}

string SPop(Stack* stack) {
    if (stack -> top == nullptr) {
        return "";
    }
    SNode* temp = stack -> top;
    string data = temp -> data;
    stack -> top = stack -> top -> next;
    stack -> size--;
    delete temp;
    return data;
}

void PrintStack(Stack* stack) {
    cout << "[";
    SNode* current = stack -> top;
    bool first = true;
    while (current != nullptr) {
        if (!first) cout << ", ";
        cout << current -> data;
        current = current -> next;
        first = false;
    }
    cout << "]" << endl;
}

void DestroyStack(Stack* stack) {
    while (stack -> top != nullptr) {
        SPop(stack);
    }
}

File: stack.go
package main

import "fmt"

type SNode struct {
	Data string
	Next *SNode
}

type Stack struct {
	Top  *SNode
	Size int
}

func CreateStack() *Stack {
	return &Stack{
		Top:  nil,
		Size: 0,
	}
}

func (stack *Stack) Push(data string) {
	newNode := &SNode{
		Data: data,
		Next: stack.Top,
	}
	stack.Top = newNode
	stack.Size++
}

func (stack *Stack) Pop() string {
	if stack.Top == nil {
		return ""
	}
	data := stack.Top.Data
	stack.Top = stack.Top.Next
	stack.Size--
	return data
}

func (stack *Stack) Print() {
	fmt.Print("[")
	current := stack.Top
	first := true
	for current != nil {
		if !first {
			fmt.Print(", ")
		}
		fmt.Print(current.Data)
		current = current.Next
		first = false
	}
	fmt.Println("]")
}